relay-sampling/src/config.rs

```diff
@@ -0,0 +1,455 @@
+}
+
+impl SamplingRule {
+    fn supported(&self) -> bool {
+        self.condition.supported() && self.ty != RuleType::Unsupported
+    }
+}
+
+/// A sampling strategy definition.
+///
+/// A sampling strategy refers to the strategy that we want to use for sampling a specific rule.
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+#[serde(tag = "type")]
+pub enum SamplingValue {
+    /// A direct sample rate to apply.
+    ///
+    /// A rule with a sample rate will be matched and the final sample rate will be computed by
+    /// multiplying its sample rate with the accumulated factors from previous rules.
+    SampleRate {
+        /// The sample rate to apply to the rule.
+        value: f64,
+    },
+
+    /// A factor to apply on a subsequently matching rule.
+    ///
+    /// A rule with a factor will be matched and the matching will continue onto the next rules
+    /// until a sample rate rule is found. The matched rule's factor will be multiplied with the
+    /// accumulated factors before moving onto the next possible match.
+    Factor {
+        /// The fator to apply on another matched sample rate.
+        value: f64,
+    },
+}
+
+impl SamplingValue {
+    pub(crate) fn value(&self) -> f64 {
+        *match self {
+            SamplingValue::SampleRate { value } => value,
+            SamplingValue::Factor { value } => value,
+        }
+    }
+}
+
+/// Defines what a dynamic sampling rule applies to.
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub enum RuleType {
+    /// A trace rule matches on the [`DynamicSamplingContext`](crate::DynamicSamplingContext) and
+    /// applies to all transactions in a trace.
+    Trace,
+    /// A transaction rule matches directly on the transaction event independent of the trace.
+    Transaction,
+    // NOTE: If you add a new `RuleType` that is not supposed to sample transactions, you need to
+    // edit the `sample_envelope` function in `EnvelopeProcessorService`.
+    /// If the sampling config contains new rule types, do not sample at all.
+    #[serde(other)]
+    Unsupported,
+}
+
+/// The identifier of a [`SamplingRule`].
+///
+/// This number must be unique within a Sentry organization, as it is recorded in outcomes and used
+/// to infer which sampling rule caused data to be dropped.
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct RuleId(pub u32);
+
+impl fmt::Display for RuleId {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// A range of time.
+///
+/// The time range should be applicable between the start time, inclusive, and
+/// end time, exclusive. There aren't any explicit checks to ensure the end
+/// time is equal to or greater than the start time; the time range isn't valid
+/// in such cases.
+#[derive(Debug, Clone, Copy, Default, Serialize, Deserialize)]
```

Comment:
Is this specific to the config module? What do you think about moving it to utils?